### PR TITLE
Add additional padding to allow expected layout of explore feed buttons

### PR DESCRIPTION
**Pull request by mcleinman**
Proposed fix for [bug T250883](https://phabricator.wikimedia.org/T250883).

**Replication steps** 
To reliable reproduce the bug on an 11" iPad Pro running iOS 13.3.1, all I had to do was turn on the `Bold Text` switch in Settings. No need to mess with the text size - I couldn't reliably reproduce the bug that way.

**Expected behavior** 
All button text (`Additional related articles`, `Save for later`, `All top read articles`, etc.) should be on one line.

**What was happening**
When iOS's settings for text is large or bolded, this text sometimes is on two lines on iPad.

----

**This is a hacky solution.**
I spent quite a while understanding our layout. I know exactly how this is happening, but I wasn't able to find the *why*. My hunch is that it's due to rounding in the various calculations, but I can't be 100% positive. **Feel free to push back if you want a more satisfying solution.** I could dig more into this if this "add 1" fix is undesirable.

**I'm doing it in all cases**
We could just add the additional 1 point when the text is large or bold. Given that the additional point has minimal UI impact, I think it's preferrable to have it always for safety. If it's a rounding error, it could easily exist in normal text in other languages (depending on character width), or some other scenario. That said, I don't feel super-strongly about this - if we just want to ad the additional point on iPad and/or when text is non-typical, let me know and I can update this PR.

**How this was happening**
When digging deep in our `extension UIButton`'s `wmf_sizeThatFits`, I could see all the calculations happening. The button image's width is calculated as 12 in this function, however when on screen that image takes a width of `13`. With the overall button width remaining constant, the button label's width is one less than expected, thus compressing the label. If you set the label as 1 line it would compress the middle part of the text; in our normal 0-line (expandable) label we'd get the 2 lines observed.

**Seemingly not an iOS bug**
Of course I Googled this problem. [An issue on a TenCent library](https://github.com/Tencent/QMUI_iOS/issues/620) is the only similar issue I found, and it has a different source cause. I tried [sizing the label](https://github.com/Tencent/QMUI_iOS/commit/2f8dc48b1503836a82fce626583008796214857a) shortly before making these calculations; it made no difference.

**Screen shots of this PR**

iPad Pro 11" w/ bold text

 

iPad Pro 11" w/ normal text

 

iPhone Xs w/ bold text

 

iPhone Xs w/ normal text

### CODE REVIEW

**Comment by staykids** [comment]
Wanted to mention during standup, I love your detective work here and great work finding this solution! I'm peeking and experimenting on it as well. If we can't find a more direct fix, I'd recommend we at least constrain the `widthAdjustment += 1` fix to just be applied to the button subclass (`AlignedImageButton`) we're experiencing the issue with so we're not accidentally causing unexpected behavior elsewhere.

**Comment by mcleinman** [comment]
> Wanted to mention during standup, I love your detective work here and great work finding this solution! I'm peeking and experimenting on it as well. If we can't find a more direct fix, I'd recommend we at least constrain the `widthAdjustment += 1` fix to just be applied to the button subclass (`AlignedImageButton`) we're experiencing the issue with so we're not accidentally causing unexpected behavior elsewhere.

Good call, thanks! I've updated this.

**Comment by mcleinman** [comment]
> I'm pretty consistently able to coax either device class (iPad/iPhone) into the two line state without your width fix, but I'm not able to identify consistent steps to get it into that state. For me, it's always been by manipulating some combination of dynamic type size and the bold text accessibility setting, but the weird thing is it's seemingly inconsistent in _reproducible_ steps to get there.
> 
> Sorry I haven't been able to discover a cleaner alternative (short of just rewriting `SaveButton`) yet. I went down a few different paths but nothing seemed to work as consistently as your solution. My only recommendation would be, if you also see value in it, to pull the fix into the `AlignedImageButton` extension below – that way it's more clear it's an `AlignedImageButton` fix and there no chance of it unintentionally affecting other `UIButton` subclasses.
> 
> ```swift
> extension AlignedImageButton {
>     public override func wmf_sizeThatFits(_ maximumSize: CGSize) -> CGSize {
>         var size = super.wmf_sizeThatFits(maximumSize)
>         // When iOS's Bold Text setting was turned on, labels on Explore feed buttons were moving to two lines,
>         // presumably due to rounding errors in this function. The extra 1 here allows them to layout as expected.
>         size.width += 1
>         return size
>     }
> }
> ```

Good call. This is now updated.

**Comment by tonisevener** [comment]
Nice investigation! I found a possible reason for this weirdness which clears a little more up but I'm still stuck on how to improve the fix. I noticed when bold text was off, view inspector correctly showed the width as 12 for `SaveButton`. But when bold was on, it jumped to 13. So it seems iOS bolds images thus slightly increasing the width of images that are set to `Template` in the asset catalog. When I changed "places-more", "saved", and "unsaved" to render as `Original`, the image stayed unbolded, which lined up with what `image.size.width` returns in `wmf_sizeThatFits` for `UIButton`.

Screen Shot 2020-05-21 at 1 55 23 PM

The problem is 1. I think we need these set as template for tinting purposes and 2. it's a good thing iOS bolds the image, I wouldn't want to lose that accessibility win. Maybe there's just a UIKit bug where it doesn't actually return the bolded image in `image(for: .normal)` when we do our size calculations. So unless we can find a way to line up that width calculation with the actual bolded image width, I think tacking on an extra 1 is the best we can do. I would just suggest maybe we wrap it up in a `if (UIAccessibilityIsBoldTextEnabled() && image != nil) {` conditional to keep it even lower impact.

**Comment by mcleinman** [comment]
> The problem is 1. I think we need these set as template for tinting purposes and 2. it's a good thing iOS bolds the image, I wouldn't want to lose that accessibility win. Maybe there's just a UIKit bug where it doesn't actually return the bolded image in `image(for: .normal)` when we do our size calculations. So unless we can find a way to line up that width calculation with the actual bolded image width, I think tacking on an extra 1 is the best we can do. I would just suggest maybe we wrap it up in a `if (UIAccessibilityIsBoldTextEnabled() && image != nil) {` conditional to keep it even lower impact.

I was hoping `image(for: [something else])` might return the `13`, but no such luck. 

I've added half of your suggestion, @tonisevener - I'm checking that an image is present, but I'm not sure we want to do this only when text is bolded. We were also seeing this behavior with larger text sizes, though inconsistently and without good repro steps. I think we should should add the additional `1` padding if there is an image present - so that it fixes bolded and different-sized text.

File: Wikipedia/Code/AlignedImageButton.swift [hunk 1]
```diff
@@ -55,6 +55,19 @@ public class AlignedImageButton: UIButton {
...
     }
+
+    public override func wmf_sizeThatFits(_ maximumSize: CGSize) -> CGSize {
+        // When iOS's Bold Text setting was turned on, labels on Explore feed buttons were moving to two lines,
+        // presumably due to rounding errors in this function. The extra 1 here allows them to layout as expected.
+        // The buttons having problems were AlignedImageButtons and SaveButtons (which are subclasses of AlignedImageButtons).
+        // (Details: The button image's width is calculated as 12, however on screen the image takes a width of 13.)
+        guard let _ = self.image(for: .normal) else {
+            // If image, it should layout fine.
+            return maximumSize
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply Wikipedia/Code/AlignedImageButton.swift]: 🤦 Good catch, thank you. Fixed this.

**Pull request merged**